Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,168 @@
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+            deviceModel.Id = Guid.NewGuid().ToString();
+
+            // Note: using UpdateAsync because the service generates the ID
+            var result = await this.storage.UpdateAsync(
+                STORAGE_COLLECTION,
+                deviceModel.Id,
+                JsonConvert.SerializeObject(deviceModel),
+                null);
+
+            deviceModel.ETag = result.ETag;
+
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
+        {
+            var id = deviceModel.Id;
+            var eTag = deviceModel.ETag;
+
+            try
+            {
+                var item = await this.GetAsync(id);
+
+                if (item.ETag == eTag)
+                {
+                    // Replace a custom device model
+                    deviceModel.Created = item.Created;
+                    deviceModel.Modified = DateTimeOffset.UtcNow;
+                    
+                    var result = await this.storage.UpdateAsync(
+                        STORAGE_COLLECTION,
+                        id,
+                        JsonConvert.SerializeObject(deviceModel),
+                        eTag);
+
+                    // Return the new ETag provided by the storage
+                    deviceModel.ETag = result.ETag;
+                }
+                else
+                {
+                    this.log.Error("Invalid ETag. Current Device Model ETag is:'", () => new { ETag = item.ETag });
+                    throw new ConflictingResourceException("Invalid ETag. Running simulation ETag is:'" + item.ETag + "'.");
+                }
+            }
+            catch (ResourceNotFoundException)
+            {
+                this.log.Info("Creating new device model via PUT.", () => { });
+
+                var result = await this.InsertAsync(deviceModel);
+                deviceModel.ETag = result.ETag;
+            }
+            catch(Exception exception)
```

Comment:
nit: space

Reply:
before this block add:
```
catch (ConflictingResourceException)
{
    throw;
}
```
so that the exception at line 141 goes out without being caught here